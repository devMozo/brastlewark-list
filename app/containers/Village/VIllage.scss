.Village {
  & > h1 {
    text-align: center;

    &:after {
      content: '';
      width: 200px;
      height: 5px;
      margin: 1rem auto 3rem;
      background-color: $orange;
      display: block;
    }
  }
}
