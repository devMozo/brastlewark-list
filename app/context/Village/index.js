import React from 'react';

export const VillageContext = React.createContext({
  citizens: [],
  loading: false,
  error: false,
  onClickCitizen: () => '',
});
