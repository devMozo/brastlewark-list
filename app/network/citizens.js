import { PROD_URL } from 'network/setting';

const CITIZEN_URL = PROD_URL + 'rrafols/mobile_test/master/data.json';

export const getCitizens = () => {
  return fetch(CITIZEN_URL, {
    method: 'GET',
    cache: 'force-cache',
  }).then(response => {
    if (response.ok) {
      return response.json();
    } else {
      throw Error('Error when tries to pick the data');
    }
  });
};
