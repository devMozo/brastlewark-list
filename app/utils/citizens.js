export const getBestWork = (citizens = []) => {
  const arrayProfessions = [];

  citizens.forEach(citizen => {
    const { professions } = citizen;

    professions &&
      professions.forEach(proffesion => {
        const index = arrayProfessions.findIndex(item => item.value === proffesion);

        if (index >= 0) {
          arrayProfessions[index] = {
            ...arrayProfessions[index],
            total: arrayProfessions[index].total + 1,
          };
        } else {
          arrayProfessions.push({
            value: proffesion,
            total: 1,
          });
        }
      });
  });

  return arrayProfessions.length === 0
    ? 0
    : arrayProfessions.reduce((prev, current) => (prev.total > current.total ? prev : current));
};

export const getAverageAge = (citizens = []) => {
  const lengthCitizens = citizens.length;
  const sumCitizensAge =
    lengthCitizens &&
    citizens.reduce((citizen, otherCitizen) => {
      return {
        age: citizen.age + otherCitizen.age,
      };
    }).age;
  return Math.round(sumCitizensAge / (lengthCitizens > 0 ? lengthCitizens : 1));
};

export const getFullCitizens = (citizens = [], friendNames = []) => {
  const fullFriends = [];

  citizens.forEach(citizen => {
    if (friendNames.includes(citizen.name)) {
      fullFriends.push(citizen);
    }
  });

  return fullFriends.length > 0 ? fullFriends : friendNames;
};
