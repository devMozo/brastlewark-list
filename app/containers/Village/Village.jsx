import React from 'react';
import { ListFeaturesVillage } from 'components/ListFeaturesVillage/ListFeaturesVillage';
import { getCitizens } from 'network/citizens';
import { CitizenLandscape } from 'components/CitizenLandscape/CitizenLandscape';
import { Loading } from 'components/Loading/Loading';
import { HandlerError } from 'components/HandlerError/HandlerError';
import { getBestWork, getAverageAge } from 'utils/citizens';

const Background = require('assets/background.jpg');

const VillageHeader = ({ title = '', citizens = [], loading, error }) => {
  const bestWork = getBestWork(citizens);
  const features = [
    {
      title: 'Total Citizens',
      value: citizens.length,
      color: 'orange',
    },
    {
      title: bestWork && bestWork.value,
      value: bestWork && bestWork.total,
      color: 'blue',
    },
    {
      title: 'Average Year',
      value: getAverageAge(citizens),
      color: 'green',
    },
  ];

  return (
    <header className="VillageHeader">
      <div className="background">
        <img src={String(Background)} />
      </div>
      <header>
        <h1 className="village__header-title padding-2-left padding-1-top font-super-cell font-color-orange">
          {' '}
          {title}'s Village{' '}
        </h1>
      </header>
      <section>
        <ListFeaturesVillage items={features} loading={loading} error={error} />
      </section>
    </header>
  );
};

const VillageCenterColumn = ({ items, loading }) => {
  return (
    <section className="VillageCenterColumn padding-2-left padding-3-right">
      {loading ? (
        <Loading />
      ) : (
        <ul>{items && items.map((item, key) => <CitizenLandscape key={key} {...item} />)}</ul>
      )}
    </section>
  );
};

const VillageRightColumn = () => {
  return (
    <section className="VillageRightColumn">
      <header>
        <img src="" />
      </header>
    </section>
  );
};

const VillageFooter = () => {
  return (
    <footer className="VillageFooter padding-2-y margin-2-top">
      <h5> by Nicolas Mozo </h5>{' '}
    </footer>
  );
};

export class Village extends React.PureComponent {
  state = {
    citizens: [],
    error: false,
    loading: true,
  };

  // To avoid the redundant setState when the component did mount
  retryGetCitizenFromServer = () => {
    this.setState({
      error: false,
      loading: true,
    });
    this.getCitizensFromServer();
  };

  getCitizensFromServer = () => {
    getCitizens()
      .then(citizens => {
        this.setState({
          citizens,
          error: false,
          loading: false,
        });
      })
      .catch(error => {
        this.setState({
          error: true,
          loading: false,
        });
      });
  };

  componentDidMount() {
    this.getCitizensFromServer();
  }

  render() {
    const { match } = this.props;
    const { citizens, error, loading } = this.state;
    const villageName = match.params.name;
    const arrayCitizens = citizens && citizens.Brastlewark;

    return (
      <section className="Village">
        <VillageHeader
          citizens={arrayCitizens}
          title={villageName}
          loading={loading}
          error={error}
        />
        <h1> {villageName}'s Citizens </h1>
        {error && (
          <HandlerError
            error="We couldn't load the data from the server"
            action={() => this.retryGetCitizenFromServer()}
          />
        )}
        {citizens && (
          <React.Fragment>
            <VillageCenterColumn loading={loading} items={arrayCitizens} />
            <VillageRightColumn />
          </React.Fragment>
        )}
        <VillageFooter />
      </section>
    );
  }
}
