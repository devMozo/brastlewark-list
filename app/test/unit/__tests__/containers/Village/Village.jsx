import { Village } from 'containers/Village/Village';
import { shallow } from 'enzyme';
import React from 'react';
import { citizens } from 'test/unit/__mocks__/citizens.json';

describe('<Village />', () => {
  let props;
  let village;
  const MOCK_NAME = 'Mocked Name';
  global.fetch = jest.fn(() => new Promise(resolve => resolve(citizens)));

  beforeEach(() => {
    props = {
      match: undefined,
    };
    village = () => {
      return shallow(<Village {...props} />);
    };
  });

  it('Snapshotesting completed <Village />', () => {
    props = {
      match: {
        params: {
          name: MOCK_NAME,
        },
      },
    };
    const mountedComponent = village();
    mountedComponent.update();

    expect(mountedComponent).toMatchSnapshot();
  });
});
