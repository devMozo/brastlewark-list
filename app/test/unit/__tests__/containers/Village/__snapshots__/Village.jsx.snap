// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Village /> Snapshotesting completed <Village /> 1`] = `ShallowWrapper {}`;
